File: nattybear/level2/타겟-넘버/solution.js
class Tree {
  constructor(n, l, r) {
    this.node = n
    this.left = l
    this.right = r
  }
}

function solution(numbers, target) {
  const depth = numbers.length
  make_tree([], depth)
  return signs
    .map(xs => zip_with((x,y) => x * y, xs, numbers))
    .map(xs => xs.reduce((acc,x) => acc + x))
    .filter(x => x == target)
    .length
}

const signs = []

function make_tree(value, depth) {
  if (depth == 0) {
    signs.push(value)
    return
  }
  const left = make_tree(value.concat(1), depth - 1)
  const right = make_tree(value.concat(-1), depth - 1)
  return new Tree(value, left, right)
}

function zip_with(f, xs, ys) {
  const zs = []
  for (let i = 0; i < xs.length; i++) {
    let x = xs[i]
    let y = ys[i]
    let z = f(x, y)
    zs.push(z)
  }
  return zs
}

File: nattybear/level2/타겟-넘버/Solution.java
import java.util.*;
import java.util.stream.*;

class Solution {
  static Stream.Builder<IntStream> signs = Stream.builder();

  public int solution(int[] numbers, int target) {
    int depth = numbers.length;
    makeTree(IntStream.empty(), depth);
    return (int) signs
      .build()
      .map(xs -> zipWith(xs, numbers))
      .mapToInt(xs -> xs.sum())
      .filter(x -> x == target)
      .count();
  }

  static IntStream zipWith(IntStream xs, int[] numbers) {
    PrimitiveIterator.OfInt ys = xs.iterator();
    IntStream.Builder zs = IntStream.builder();
    for (int i = 0; i < numbers.length; i++) {
      int n = numbers[i];
      int y = ys.nextInt();
      int z = n * y;
      zs.add(z);
    }
    return zs.build();
  }

  static Tree makeTree(IntStream value, int depth) {
    if (depth == 0) {
      signs.add(value);
      return new Tree();
    }
    int[] xs = value.toArray();
    Tree left = makeTree(
      IntStream.concat(Arrays.stream(xs), IntStream.of(1)), depth - 1);
    Tree right = makeTree(
      IntStream.concat(Arrays.stream(xs), IntStream.of(-1)), depth - 1);
    return new Tree(value, left, right);
  }
}

class Tree {
  IntStream node;
  Tree left;
  Tree right;

  Tree() {}

  Tree(IntStream n, Tree l, Tree r) {
    node = n;
    left = l;
    right = r;
  }
}
